pep-perennial.rst

```diff
@@ -0,0 +1,161 @@
+Last-Modified: $Date$
+Author: Nathaniel J. Smith <njs@pobox.com>
+        Thomas Kluyver <thomas@kluyver.me.uk>
+BDFL-Delegate: Nick Coghlan <ncoghlan@gmail.com>
+Discussions-To: Distutils SIG <distutils-sig@python.org>
+Status: Active
+Type: Informational
+Content-Type: text/x-rst
+Created: 3-May-2019
+Post-History: 3-May-2019
+
+Abstract
+========
+
+This PEP proposes a scheme for new 'manylinux' distribution tags to be defined
+without requiring a PEP for every specific tag. The naming scheme is based on
+glibc versions, with profiles in the auditwheel tool defining what other
+external libraries and symbols a compatible wheel may link against.
+
+While there is interest in defining tags for non-glibc Linux platforms,
+this PEP does not attempt to address that.
+
+Rationale
+=========
+
+Distributing compiled code for Linux is more complicated than for other popular
+operating systems, because the Linux kernel and the libraries which typically
+accompany it are built in different configurations and combinations for different
+distributions. However, there are certain core libraries which can be expected in
+many common distributions and are reliably backwards compatible, so binaries
+built with older versions of these libraries will work with newer versions.
+:pep:`513` describes these ideas in much more detail.
+
+The ``manylinux1`` (:pep:`513`) and ``manylinux2010`` (:pep:`571`) tags make
+use of these features. They define a set of core libraries and symbol versions
+which wheels may expect on the system, based on CentOS 5 and 6 respectively.
+Typically, packages are built in Docker containers based on these old CentOS
+versions, and then the ``auditwheel`` tool is used to check them and bundle any
+other linked libraries into the wheel.
+
+If we were to define a ``manylinux2014`` tag based on CentOS 7, there would be
+five steps involved to make it practically useful:
+
+1. Write a PEP
+2. Prepare docker images based on CentOS 7.
+3. Add the definition to auditwheel
+4. Allow uploads with the new tag on PyPI
+5. Add code to pip to recognise the new tag and check if the platform is
+   compatible
+
+Although preparing the docker images and updating auditwheel take more work,
+these parts can be used as soon as that work is complete. The changes to pip
+are more straightforward, but not all users will promptly install a new version
+of pip, so distributors are concerned about moving to a new tag too quickly.
+
+This PEP aims to remove the need for steps 1 and 5 above, so new manylinux tags
+can be adopted more easily.
+
+Naming
+======
+
+Tags using the new scheme will look like::
+
+    manylinux_glibc_2_17_x86_64
+
+Where ``2_17`` is the major and minor version of glibc. I.e. for this example,
+the platform must have glibc 2.17 or newer. Installer tools should be prepared
+to handle any numeric values here, but building and publishing wheels to PyPI
+will probably be constrained to specific profiles defined by auditwheel.
```

Comment:
I guess the counterargument is that `manylinux_17` doesn't have much clear meaning, especially in the context of having already had `manylinux1` and `manylinux2010`.

Leaving out the `2_` also doesn't give us an obvious future numbering scheme if glibc 3.x does turn out one day, whether that's a compatibility-breaking release like Python 3 or a continuation more like Linux 3.

I don't have strong feelings over this, and I'm assuming you've thought more about this whole space than I have. But maybe if it needs further discussion we should take that somewhere where more people will participate. (There's nothing like more people for a naming discussion!)

Reply:
It's awkward right now, but in a few years `manylinux1` and `manylinux2010` will be gone and forgotten. And the problem with mentioning `glibc` explicitly is that several people thought it had a clear meaning, but were wrong about what that meaning was :-).

`manylinux_2_17_x86_64` isn't terrible either. But to explain why I'm not worried about glibc 3: Well first, glibc will definitely never pull a Linux 3, because the version corresponds to the soname. A glibc 3 would effectively be a totally new project with absolutely no binary compatibility; it would have no more connection to glibc 2 than it would to musl or dietlibc.

In theory it's possible that the glibc devs would decide to fork off a new project, call it glibc 3, and that all the distros would decide to switch to it. But they've been maintaining ABI compatibility for 22 years now; presumably if the ABI had a fundamental unfixable problem they'd have noticed by now. So it's kind of like preparing for a meteorite strike: sure, it *could* happen, but... something else will get you first :-). I think it's at least as plausible that all the distros will switch to musl, or that Linux will get replaced by Fuchsia, or x86-64 will get replaced by RISC-V, than that glibc 3 will appear and take over the world. And if it did it would still take like a decade to transition, so we'd have plenty of time to figure out what to do.

If we want to be really thorough with our due-diligence we could contact the glibc devs and ask them to explicitly confirm that glibc 3 is never going to happen.

Reply:
As a part-time glibc developer, I can tell you that `libc.so.7` will _almost surely_ never happen; however, there has been some discussion of a version 3.0 of the glibc source distribution.  It would still build `libc.so.6` but with some of the very oldest compatibility symbols dropped (relating to things like the attempt to make stdio `FILE` be the same thing as libstdc++ `filebuf`, which was abandoned as unworkable more than 20 years ago, but still complicates the stdio implementation).  This is not likely to happen anytime soon.

Reply:
@zackw Interesting.

The challenge with a 3.0 source version is that we basically assume that can check the "ABI compatibility level" of glibc by calling `gnu_get_libc_version`, and that this will return a string like `"2.${ABI version}${arbitrary vendor-specific junk}"`, where `${ABI version}` is an integer. Then we do automated things with `${ABI version}`.

(And note that this is orthogonal to the "perennial" part of this proposal – a hypothetical glibc 3.0 release would also break manylinux1, manylinux2010, manylinux2014, etc.)

Anyway, in addition to breaking on a hypothetical future glibc 3.0, the current string-parsing hack is kinda gross, and means the glibc devs have API surface area that they don't know about. Maybe this means that they can never use 3.0 as a release name, because it would break all existing manylinux wheels. It's awkward all around.

I suspect the right solution is for glibc to add a new API, like `gnu_get_libc_abi_level`, which returns an integer. This integer has the following properties:

- for old glibc 2.X versions, the ABI level is retroactively defined to equal the minor version number, so we can fall back on `gnu_get_libc_version` + string parsing
- code built against a glibc with ABI level X will also work when run against a glibc with ABI level Y, so long as X < Y

That lets us (eventually) drop our gross version parsing regex, makes the actual guarantees more visible to the glibc devs, and potentially provides a path to releasing 3.0 without breaking everyone (`gnu_get_libc_version` can return `"3.0"` but `gnu_get_libc_abi_level` would keep incrementing monotonically). And then the perennial manylinux tag name would be defined as `manylinux_${glibc ABI level}_${platform}`. (Which is conceptually different from the `manylinux_${glibc minor version number}_${platform}` I suggested above, but identical in practice.)

Does that sound sensible to you?

If we wanted to bring this to the glibc devs, how would we do that? (And is there any way I can do it without signing up for the glibc dev firehose?)

Reply:
This sounds sensible in outline, except that your second property 

> * code built against a glibc with ABI level X will also work when run against a glibc with ABI level Y, so long as X <= Y

would _not_ be guaranteed for the hypothetical 3.0.  There would be another number, Z, indicating the oldest supported ABI level; code built against level X would work as long as Z <= X <= Y.  (Currently Z is 0 and the hypothetical 3.0 would increase it to, um, probably 3 or 4.)  Z could also be exposed via a function call, perhaps `gnu_get_libc_min_abi_level`.

I think the best way to bring this to the attention of the glibc devs is to file a wishlist bug report at https://sourceware.org/bugzilla/ .

Reply:
> I think the best way to bring this to the attention of the glibc devs is to file a wishlist bug report at https://sourceware.org/bugzilla/ .

https://sourceware.org/bugzilla/show_bug.cgi?id=24636

Reply:
Bringing over my preference from the Discourse thread: my view is the opposite of Nathaniel's, and would favour the slightly longer `manylinux_bp_glibc_2_17_x86_64`. The intent of the `_bp` addition is that it makes the first part stand for "manylinux build profile", and then then rest of the string names the exact build profile. If we're worried about filename length, then we could abbreviate it further to `mlbp_glibc_2_17_x86_64`, but that has the downside of losing the useful property of having "linux" as a visible part of the filename.

Reply:
Based on the discussion in that glibc issue, it sounds like the recommendation is to assume that glibc 3.x might happen, and to assume that the major version bump *won't* indicate a compatibility break. So I'm changing my recommendation here to `manylinux_2_17_x86_64`.

Also, apparently, we should be going back and revising the previous manylinux PEPs and pip code to assume that glibc 3 is compatible with glibc 2. I guess the simplest way to handle that would be to make this PEP say the right thing, and then say `manylinux1` and `manylinux2010` are redefined as aliases for `manylinux_2_5` and `manylinux_2_12`, with the new very-slightly-different semantics.

I don't see any value in sticking `_bp_glibc_` in there. `manylinux` and `manylinux_bp_glibc` mean exactly the same things, except one has extra clutter. What could any manylinux tag even be besides a "build profile"?

Reply:
With that, I'd expect to be able to go look up a spec called "manylinux 2.17". There's no such spec - there's a manylinux 2014 spec that calls out a glibc 2.17 build profile.

So if we're going to switch from naming the spec version in the filename to instead naming the build profile, the filename should reflect that.

Reply:
> there's a manylinux 2014 spec

Sure, that's one way we could do things, but... that's not what this PEP is proposing.

> I'd expect to be able to go look up a spec called "manylinux 2.17"

The spec you would want is the "manylinux" spec, which seems reasonable. It's exactly the same as how when you see a wheel tagged `macos_10_10`, you don't look for the "macos 10.10" spec, you look for the "macos" spec. (Well, actually it's much harder to figure out the relevant spec than that – you have to figure out that it's [from PEP 425](https://www.python.org/dev/peps/pep-0425/#platform-tag), which incorporates the behavior of `distutils.util.get_platform()` by reference, and then figure out in what circumstances `distutils` returns strings that start with macos.)